src/main/java/com/questhelper/steps/DetailedQuestStep.java

```diff
@@ -392,49 +391,41 @@ public void onItemDespawned(ItemDespawned itemDespawned)
 
 		if (tileHighlights.containsKey(tile))
 		{
-			tileHighlights.get(tile).remove(Integer.valueOf(itemDespawned.getItem().getId()));
+			tileHighlights.get(tile).remove(itemDespawned.getItem().getId());
 		}
 	}
 
-	private void addItemTiles()
+	protected void addItemTiles(Collection<Requirement> requirements)
 	{
 		if (requirements == null)
 		{
 			return;
 		}
 		Tile[][] squareOfTiles = client.getScene().getTiles()[client.getPlane()];
 
-		for (Tile[] lineOfTiles : squareOfTiles)
-		{
-			for (Tile tile : lineOfTiles)
-			{
-				if (tile == null)
-				{
-					continue;
-				}
-				List<TileItem> items = tile.getGroundItems();
-				if (items != null)
-				{
-					for (TileItem item : items)
-					{
-						for (Requirement requirement : requirements)
-						{
-							if (requirement != null && requirement.getClass() == ItemRequirement.class
-								&& ((ItemRequirement) requirement).isActualItem()
-								&& ((ItemRequirement) requirement).getAllIds().contains(item.getId()))
-							{
-								tileHighlights.computeIfAbsent(tile, k -> new ArrayList<>());
-								tileHighlights.get(tile).add(item.getId());
-								break;
-							}
-						}
-					}
-				}
-			}
-		}
+		Stream.of(squareOfTiles)
+			.flatMap(Arrays::stream) // lineOfTiles
+			.filter(Objects::nonNull)
+			// map to multimap
+			.map(tile -> Multimaps.index(tile.getGroundItems() == null ? new ArrayList<>() : tile.getGroundItems(), k -> tile))
+			// map to stream of map entries
+			.flatMap(it -> it.entries().stream().map(e -> new AbstractMap.SimpleEntry<>(e.getKey(),e.getValue())))
+			// collect to a map of <Tile, List<TileItem>>
+			.collect(Collectors.groupingBy(Map.Entry::getKey, Collectors.mapping(Map.Entry::getValue, Collectors.toList())))
+			.forEach((tile, itemList) -> {
+				itemList.forEach(i -> {
+					requirements.stream()
+						.filter(Objects::nonNull)
+						.filter(req -> req.getClass() == ItemRequirement.class)
+						.map(ItemRequirement.class::cast)
+						.filter(ItemRequirement::isActualItem)
+						.filter(req -> req.getAllIds().contains(i.getId()))
+						.forEach(req -> tileHighlights.get(tile).add(i.getId()));
+				});
+			});
```

Comment:
So this does work. I opened 20-30 quests and received no errors.
If you have questions about it let me know. If you think it's too much then I can revert it.

Reply:
I think this one is getting a bit much, I think there's better ways for us to break it down and make it cleaner. Somehow the filters are about the same number of lines excluding brackets and the same depth 😅 .